### PR TITLE
Convert RealWebSocket to Kotlin

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.kt [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+  private val maskKey: ByteArray? = if (isClient) ByteArray(4) else null
+  private val maskCursor: Buffer.UnsafeCursor? = if (isClient) Buffer.UnsafeCursor() else null
+
+  /** Send a ping with the supplied [payload]. */
+  @Throws(IOException::class)
+  fun writePing(payload: ByteString) {
+    writeControlFrame(OPCODE_CONTROL_PING, payload)
+  }
+
+  /** Send a pong with the supplied [payload]. */
```
Comment (jemaystermind) [inline okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.kt]:
```suggestion
  /** Send a ping with the supplied [payload]. */
```

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt [hunk 2]
```diff
@@ -13,430 +13,403 @@
...
-    this.listener = listener;
-    this.random = random;
-    this.pingIntervalMillis = pingIntervalMillis;
 
-    byte[] nonce = new byte[16];
-    random.nextBytes(nonce);
-    this.key = ByteString.of(nonce).base64();
+    val nonce = ByteArray(16)
+    random.nextBytes(nonce)
+    this.key = ByteString.of(*nonce).base64()
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: You could inline this with something like

`this.key = ByteArray(16).apply { random.nextBytes(this) }.toByteString().base64()`

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt [hunk 3]
```diff
@@ -13,430 +13,403 @@
...
+      throw ProtocolException(
+          "Expected 'Connection' header value 'Upgrade' but was '$headerConnection'")
     }
 
-    String headerUpgrade = response.header("Upgrade");
-    if (!"websocket".equalsIgnoreCase(headerUpgrade)) {
-      throw new ProtocolException(
-          "Expected 'Upgrade' header value 'websocket' but was '" + headerUpgrade + "'");
+    val headerUpgrade = response.header("Upgrade")
+    if (!"websocket".equals(headerUpgrade, ignoreCase = true)) {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: a custom check method seems warranted

`protocolCheck(boolean) { "message" }`

File: okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.kt [hunk 4]
```diff
@@ -13,223 +13,225 @@
...
+    frameLength = (b1 and B1_MASK_LENGTH).toLong()
+    if (frameLength == PAYLOAD_SHORT.toLong()) {
+      frameLength = (source.readShort() and 0xffff).toLong() // Value is unsigned.
+    } else if (frameLength == PAYLOAD_LONG.toLong()) {
+      frameLength = source.readLong()
       if (frameLength < 0) {
-        throw new ProtocolException(
-            "Frame length 0x" + Long.toHexString(frameLength) + " > 0x7FFFFFFFFFFFFFFF");
+        throw ProtocolException(
+            "Frame length 0x${java.lang.Long.toHexString(frameLength)} > 0x7FFFFFFFFFFFFFFF")
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.kt]: Weird that there isn't a nice form of this that handles unsigned nums.  Guess that's coming.

File: okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.kt [hunk 5]
```diff
@@ -13,245 +13,246 @@
...
+  fun writeClose(code: Int, reason: ByteString?) {
+    var payload = ByteString.EMPTY
     if (code != 0 || reason != null) {
       if (code != 0) {
-        validateCloseCode(code);
+        validateCloseCode(code)
       }
-      Buffer buffer = new Buffer();
-      buffer.writeShort(code);
+      val buffer = Buffer()
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.kt]: good candidate for Buffer().apply {}

File: okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.kt (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.kt]: nope, we’re sending a pong here.

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: Done. I still want a random byte string thing. Maybe just `ByteString.random(Random(), 16)` should work.

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: From 4 years ago! https://github.com/square/okio/issues/132

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: Yeah, agreed. I wanna punt to follow-up though.

File: okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.kt (hunk 4 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.kt]: the unsigned bit is the gotcha. It’s tempting to do Long.toString(16) but it doesn’t do what we want.

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/ws/WebSocketReader.kt]: (I guess `toUlong().toString(16)` will be semantic)

File: okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.kt (hunk 5 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/ws/WebSocketWriter.kt]:
```
      payload = Buffer().run {
        writeShort(code)
        if (reason != null) {
          write(reason)
        }
        readByteString()
      }
```

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+    messageAndCloseQueue.add(Close(code, reasonBytes, cancelAfterCloseMillis))
+    runWriter()
+    return true
+  }
+
+  private fun runWriter() {
+    assert(Thread.holdsLock(this))
+
+    if (executor != null) {
+      executor!!.execute(writerRunnable)
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: Could be safe-call operator

File: okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt [hunk 7]
```diff
@@ -0,0 +1,579 @@
...
+
+  fun failWebSocket(e: Exception, response: Response?) {
+    val streamsToClose: Streams?
+    synchronized(this) {
+      if (failed) return // Already failed.
+      failed = true
+      streamsToClose = this.streams
+      this.streams = null
+      if (cancelFuture != null) cancelFuture!!.cancel(false)
+      if (executor != null) executor!!.shutdown()
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: Both could use safe-call operators

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/ws/RealWebSocket.kt]: https://github.com/square/okhttp/pull/5073

**Pull request merged**